modules/@apostrophecms/doc/lib/legacy-migrations.js

```diff
@@ -0,0 +1,181 @@
+// Migrations relevant only to those who used early alpha and beta versions of 3.x
+// are kept here for tidiness
+
+const _ = require('lodash');
+
+module.exports = (self) => {
+  return {
+    addLegacyMigrations() {
+      self.addDuplicateOrMissingWidgetIdMigration();
+      self.addDraftPublishedMigration();
+      self.addLastPublishedToAllDraftsMigration();
+      self.addLastPublishedToAllPublishedDocsMigration();
+      self.addAposModeMigration();
+      self.addStoreRelationshipIdsAsAposDocIdsMigration();
+    },
+    addDuplicateOrMissingWidgetIdMigration() {
+      self.apos.migration.add('duplicate-or-missing-widget-id', async () => {
+        return self.apos.migration.eachDoc({}, 5, async (doc) => {
+          const widgetIds = {};
+          const patches = {};
+          // Walk the areas in a doc. Your iterator function is invoked once
+          // for each area found, and receives the
+          // area object and the dot-notation path to that object.
+          // note that areas can be deeply nested in docs via
+          // array schemas.
+          //
+          // If the iterator explicitly returns `false`, the area
+          // is *removed* from the page object, otherwise no
+          // modifications are made. This happens in memory only;
+          // the database is not modified.
+          self.apos.area.walk(doc, (area, dotPath) => {
+            if (!area.items) {
+              return;
+            }
+            for (let i = 0; (i < area.items.length); i++) {
+              const item = area.items[i];
+              if (!item) {
+                continue;
+              }
+              if ((!item._id) || _.has(widgetIds, item._id)) {
```

Comment:
This seems like an easy replacement to get rid of the lodash import.

Reply:
I added a changelog entry.

`_.has` is omitted from most "you don't need lodash anymore" articles because it's actually still a bit goofy to replace.  I guess `Object.keys(foo).includes('bar')` would do it, but my goal in this work was to relocate this code with no risk of functional changes at minimum cost and the simplest way to ensure that is not to make unnecessary changes to the logic, I think.

Reply:
I'm speaking of this particular case. It's only one level deep, so it could be replaced with `widgetIds && widgetIds[item._id]`

Reply:
I'm speaking of this particular case. It's only one level deep, so it could be replaced with `widgetIds && widgetIds[item._id]`. Am I missing something there?

Reply:
I removed it. We made widgetIds so we don't have to test whether it exists. The ids came from Apostrophe so we don't have to worry about a sneaky one called `hasOwnProperty`, therefore we can do the lazy test with `widgetIds[item._id]`.